### DISCUSSION TITLE
How to configure tls for a loadbalancer setup

### DISCUSSION THREAD

**Discussion by spoorthyredi2** [strimzi/strimzi-kafka-operator#7585] [Q&A]
Hi ,

I have a working strimzi Kafka with a plaintext loadbalancer listener setup on GKE. Now, I want to setup TLS. I am quite new to the certs and secrets space. I have couple of questions.

1. I see that when I enable tls=true , I see a cert getting generated in my Kafka deployment. What is this cert ? How and where do I use it ?  Where do I get my Keys and ca.crt ?
Is there a video demonstration that can make the setup easy . I went through the Strimzi doc but am not quite clear yet.

2. How to connect to Kafka with TLS from my microservice applications ?
Currently we use just IP:Port

3. Is Kafka connect mandatory in this scenario ?
Appreciate any help .

**Comment by scholzj** [comment]
Re 1) Did you checked the docs? I'm pretty sure it is covered there. By default, when you set `tls: true`, it will use Strimzi's own certificate. You can get the public key from the secret names `<clusterName>-cluster-ca-cert` and use it in your clients. If you want, you can also provide your own server certificate which will be used by Kafka (search the docs for _custom listener certificate_).

Re 2) You will need to configure the security protocol and provide the public key of the Strimzi CA (or of the CA which signed your custom listener certificate if you decide to use it) as mentioned above.

Re 3) Kafka Connect is a framework for running Kafka connectors -> connectors are Kafka plugins which help it to integrate with other systems. It is not mandatory to use it. But it might be useful if you integrate your Kafka with some other systems (things such as other messaging systems, databases, SAP, storage such as Amazon S3 etc.). But if you just want to use Kafka from your microservices, you might not need it.

**Reply by spoorthyredi2** [reply]
ok .. do I need to create a User in order to access Kafka with tls ?

**Reply by scholzj** [reply]
The `tls: true` enables TLS encryption (and TLS server authentication). So no, for that, you do not need any user and you do not need to configure any keystore in the clients.

If you want to use TLS client authentication as well, you have to enable it using:
```yaml
authentication:
  type: tls
```
Then you would need to configure KafkaUser as well and configure keystore / private keys in your clients.

**Reply by spoorthyredi2** [reply]
We need just the Server authentication. 

However, all references out there on strimzi or other articles, are for mutual TLS.  Can you point me to docs on how to setup kafka clients on kubernetes to use basic tls encryption. Thanks in advance !!

**Reply by scholzj** [reply]
You just configure the protocol to `SSL` and configure the truststore. It is subset of what you do for mTLS -> you just kip the keystore part.

**Reply by spoorthyredi2** [reply]
These are the commands i used  for sample tests without tls.  what is the equivalent of with tls ... 

 kubectl -n strimzi-kafka run kafka-producer -ti --image=quay.io/strimzi/kafka:0.31.1-kafka-3.2.3 --rm=true --restart=Never -- bin/kafka-console-producer.sh --bootstrap-server x.x.x.x:9099 --topic my-topic
kubectl -n strimzi-kafka run kafka-consumer -ti --image=quay.io/strimzi/kafka:0.31.1-kafka-3.2.3 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server x.x.x.x:9099 --topic my-topic --from-beginning

**Reply by spoorthyredi2** [reply]
There are 4 default certs - 2 clients and 2 cluster - that got created by kafka. 
Im assuming I have to use the clients ones.

so they have  ca.key, ca.crt, ca.p12, ca.password .  Should I import these to a new truststore ?

![image](https://user-images.githubusercontent.com/116352972/200487579-8dab02aa-7102-4967-8466-6cd5f85f51f7.png)

**Reply by scholzj** [reply]
I told you right at the beginning which secret should you use for the truststore. It is up to you if you pick the `ca.crt` and create a new truststore with it or if you pick the `ca.p12` and use it as the truststore.

**Comment by spoorthyredi2** [comment]
Thanks for your help, I have pushed the ca.crt to my cacerts and i have a properties file in my local working directory.

And iam running my earlier command using props file like -- 
![image](https://user-images.githubusercontent.com/116352972/200518306-f8c9c70b-1188-431b-9e48-1088278c8dee.png)

Is it the right way or should I put my properties file inside my container.

**Reply by spoorthyredi2** [reply]
this is my properties file 
![image](https://user-images.githubusercontent.com/116352972/200518543-a21d03d7-b0b2-4b28-9c40-dfb60373c229.png)

**Reply by scholzj** [reply]
That is not how it works. You cannot create a local file and reference it from inside some container inside Kubernetes just like that. You created load balancer listener to be able to access the cluster from the outside of Kubernetes. So just connect to it from your local computer without any containers.

**Comment by spoorthyredi2** [comment]
Our loadbalancer setup is to be able to expose Kafka for use by Airflow/Astronomer.
Before apps use kafka, I had tested my _non-tls_ deployment using the below commands. 

kubectl -n strimzi-kafka run kafka-producer -ti --image=quay.io/strimzi/kafka:0.31.1-kafka-3.2.3 --rm=true --restart=Never -- bin/kafka-console-producer.sh --bootstrap-server x.x.x.x:9099 --topic my-topic
kubectl -n strimzi-kafka run kafka-consumer -ti --image=quay.io/strimzi/kafka:0.31.1-kafka-3.2.3 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server x.x.x.x:9099 --topic my-topic --from-beginning

Hence , wanted to test the tls deployment too. Is there a way I can test using above commands ?

**Reply by scholzj** [reply]
I don't think it matters that you want it to be used by Airflow / Astronomer. I think you can still use the listener now to test if it works, or? No, those commands cannot be used for that. You would need to run them differently to be able to copy files into the containers etc.

**Reply by spoorthyredi2** [reply]
Nope , after I enabled tls, I am not able to test using the same commands as I think it expects the certs. as per the docs I need to use client.properties - but I guess thats with standalone kafka clients, not containers.

**Reply by spoorthyredi2** [reply]
Finallly !!!!!! Tls is working . I dint know I had to use Kafka cli :P ... I was just trying with sample images and then created custom images with ssl props file and all. But it was quite simple with CLI .

Thanks so much @scholzj  . Appreciate your quick prompt response all the while. Cheers !!!

**Reply by scholzj** [reply]
Great, glad it works now 👍
